### PR TITLE
Float funque plus features

### PR DESCRIPTION
**Pull request by MallikarjunKamble**
**Floating Point Implementation for new features:**

- Added support for 5-Tap nadeanu spatial csf
- Added support for nadenau_weight, li, hill, watson and mannos_weight wavelet csf
- Added MS_ESSIM feature
- Added ST-RRED feature

**Fixed:**

- Repeated aligned_malloc memory allocation for dwt buffers has been removed
- Added json file for float_funque
- Added features list in integer_funque

### CODE REVIEW

File: libvmaf/src/feature/third_party/funque/funque_ssim.c [hunk 1]
```diff
@@ -93,5 +93,180 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
...
+#endif
+            index_cum += 2;
+        }
+        index_cum += (width * 2);
+    }
+
+#if ENABLE_MINK3POOL
+    double ssim_val = 1 - cbrt((double) cube_1minus_map / (width * height));
+    score->ssim_mean = ssim_clip(ssim_val, 0, 1);
+#else
```
Comment (cosmin) [inline libvmaf/src/feature/third_party/funque/funque_ssim.c]: I think we can drop all the CoV pooling code and stick with mink pooled only as `essim` and `ms-essim` and simultaneously compute mean pooled values as `ssim` and `ms-ssim`.

File: libvmaf/src/feature/third_party/funque/x86/hbd_resizer_avx2.c [hunk 2]
```diff
@@ -346,6 +345,7 @@ void hbd_vresize_avx2(const int **src, unsigned short *dst, const short *beta, i
         __m256i accum_23_0 = _mm256_add_epi64(mul2_0, mul3_0);
         __m256i accum_01_4 = _mm256_add_epi64(mul0_4, mul1_4);
         __m256i accum_23_4 = _mm256_add_epi64(mul2_4, mul3_4);
+
```
Comment (cosmin) [inline libvmaf/src/feature/third_party/funque/x86/hbd_resizer_avx2.c:348]: can we clean up these unrelated whitespace changes from this pull request and leave only the actually modified files?

**Comment by MallikarjunKamble** [comment]
**Fixed Point Implementation:**

- Added support for 5-Tap nadeanu spatial csf
- Added support for wavelet csf (nadenau weights)
- Added MS_SSIM integer feature
- Added ST-RRED integer feature

**Enhancements to Floating point:**

- Memory optimizations in MS_SSIM
- Tested floating point code to mactch across all 4 levels. Added fixes to match the python code.
-  - Fixes added in MS_SSIM and STRRED float code to match all 4 levels

File: libvmaf/src/feature/third_party/funque/funque_filters.c [hunk 3]
```diff
@@ -132,58 +131,63 @@ void spatial_filter(float *src, float *dst, int width, int height)
...
         -0.01890698454168517061991 ,
         -0.01608514932055564797264 ,
-        -0.01373463642215844680849
-    };
-    
+        -0.01373463642215844680849 };
+
+void spatial_csfs(float *src, float *dst, int width, int height, float *tmp_buf, int num_taps)
+{
+    int fwidth;
```
Comment (abhinaukumar) [inline libvmaf/src/feature/third_party/funque/funque_filters.c:138]: `fwidth` appears to be a duplicate of `num_taps`.

File: libvmaf/src/feature/third_party/funque/integer_funque_filters.c [hunk 4]
```diff
@@ -302,18 +319,27 @@ static inline void integer_horizontal_filter(spat_fil_inter_dtype *tmp, spat_fil
...
-        14498, 4677, -66, -1547, -1798, -1669, -1452, -1239, -1054, -900
-    };
+const spat_fil_coeff_dtype i_nadeanu_filter_coeffs[5] = {1658, 15139, 31193, 15139, 1658};
 
-    int src_px_stride = width;
-    int dst_px_stride = width;
+void integer_spatial_filter(void *src, spat_fil_output_dtype *dst, int dst_stride, int width,
+                            int height, int bitdepth, spat_fil_inter_dtype *tmp, int num_taps)
+{
+    int fwidth;
```
Comment (abhinaukumar) [inline libvmaf/src/feature/third_party/funque/integer_funque_filters.c:331]: `fwidth` appears to be a duplicate of `num_taps`.

File: libvmaf/src/feature/third_party/funque/funque_ssim.c [hunk 5]
```diff
@@ -93,5 +93,149 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
...
+            ssim = l * cs;
+#if ENABLE_MINK3POOL
+            cube_1minus_map += pow((1 - (l * cs)), 3);
+#else
+            ssim_sum += (l * cs);
+            l_sum += l;
+            cs_sum += cs;
+            ssim_sq_sum += (l * cs) * (l * cs);
+            l_sq_sum += l * l;
+            cs_sq_sum += cs * cs;
```
Comment (abhinaukumar) [inline libvmaf/src/feature/third_party/funque/funque_ssim.c:171]: `l` and `cs` need to be mink3 pooled for use in MS-ESSIM.

File: libvmaf/src/feature/third_party/funque/funque_ssim.c [hunk 6]
```diff
@@ -93,5 +93,149 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
...
+    score->l_cov = l_cov;
+    score->cs_cov = cs_cov;
+#endif
+
+    ret = 0;
+
+    return ret;
+}
+
+int compute_ms_ssim_mean_scales(MsSsimScore* score, int n_levels)
```
Comment (abhinaukumar) [inline libvmaf/src/feature/third_party/funque/funque_ssim.c:218]: This function doesn't appear to be handled appropriately under `ENABLE_MINK3POOL`. Flagging to double-check.

File: libvmaf/src/output.c [hunk 7]
```diff
@@ -99,7 +99,7 @@ int vmaf_write_output_xml(VmafContext *vmaf, VmafFeatureCollector *fc,
             int err = vmaf_feature_score_pooled(vmaf, feature_name, j, &score,
                                                 0, pic_cnt - 1);
             if (!err)
-                fprintf(outfile, "%s=\"%.6f\" ", pool_method_name[j], score);
+                fprintf(outfile, "%s=\"%.30f\" ", pool_method_name[j], score);
```
Comment (abhinaukumar) [inline libvmaf/src/output.c:102]: Are these `.30f` formattings deliberately left in or are they holdovers from debugging?

**Comment by MallikarjunKamble** [comment]
**Fixed Point Implementation:**

- Addition of Multi-Scale feature computation (All 4 scales)
- Addition of support for crop width and crop height
- Addition of MINK3 pool in MS_SSIM
- Bug fixes for accuracy improvements wrt python

**Enhancements to Floating point:**

- Addition of support for crop width and crop height
- Addition of MINK3 pool in MS_SSIM
- Addition of filter coeffs for Li and Hill

File: libvmaf/src/feature/third_party/funque/funque_ssim.c (hunk 1 above)
Comment (MallikarjunKamble) [reply libvmaf/src/feature/third_party/funque/funque_ssim.c]: Mink3 pooling is added. Covariance based pooling can be removed after review

File: libvmaf/src/feature/third_party/funque/funque_filters.c (hunk 3 above)
Comment (MallikarjunKamble) [reply libvmaf/src/feature/third_party/funque/funque_filters.c:138]: We will address num_taps redundancy in the next commit

File: libvmaf/src/feature/third_party/funque/integer_funque_filters.c (hunk 4 above)
Comment (MallikarjunKamble) [reply libvmaf/src/feature/third_party/funque/integer_funque_filters.c:331]: We will address num_taps redundancy in the next commit

File: libvmaf/src/feature/third_party/funque/funque_ssim.c (hunk 5 above)
Comment (MallikarjunKamble) [reply libvmaf/src/feature/third_party/funque/funque_ssim.c:171]: Addressed. Please review and close.

File: libvmaf/src/feature/third_party/funque/funque_ssim.c (hunk 6 above)
Comment (MallikarjunKamble) [reply libvmaf/src/feature/third_party/funque/funque_ssim.c:218]: This is ssim. We will correct the code in ssim and add support for mink3 pooling and update in next commit.

File: libvmaf/src/output.c (hunk 7 above)
Comment (MallikarjunKamble) [reply libvmaf/src/output.c:102]: This is to take care of ST-RRED accuracy. ST-RRED values go upto E-30
